### Exploratory Data Analysis

Exploratory data analysis (EDA) is used by data scientists to analyze and investigate data sets and summarize their main characteristics, often employing data visualization methods. It helps determine how best to manipulate data sources to get the answers you need

#### Booking Sports Facilities Dataset

In this Assessment test we will use database with 3 multiple related tables

[facilities.csv](https://github.com/tinashdj/Booking-Sports-Facilities/blob/main/datasets/facilities.csv),
[members.csv](https://github.com/tinashdj/Booking-Sports-Facilities/blob/main/datasets/members.csv),
[booking.csv](https://github.com/tinashdj/Booking-Sports-Facilities/blob/main/datasets/booking.csv).

#### In this assesment we have 14 questions to solve

### 1. How can you retrieve all the information from the cd.facilities table?

Retrieve all the information from the cd.facilities table.

In [6]:
%%sql 

SELECT * from cd.facilities

 * postgresql://postgres:***@localhost/exercises
9 rows affected.


[(0, 'Tennis Court 1', Decimal('5'), Decimal('25'), Decimal('10000'), Decimal('200')),
 (1, 'Tennis Court 2', Decimal('5'), Decimal('25'), Decimal('8000'), Decimal('200')),
 (2, 'Badminton Court', Decimal('0'), Decimal('15.5'), Decimal('4000'), Decimal('50')),
 (3, 'Table Tennis', Decimal('0'), Decimal('5'), Decimal('320'), Decimal('10')),
 (4, 'Massage Room 1', Decimal('35'), Decimal('80'), Decimal('4000'), Decimal('3000')),
 (5, 'Massage Room 2', Decimal('35'), Decimal('80'), Decimal('4000'), Decimal('3000')),
 (6, 'Squash Court', Decimal('3.5'), Decimal('17.5'), Decimal('5000'), Decimal('80')),
 (7, 'Snooker Table', Decimal('0'), Decimal('5'), Decimal('450'), Decimal('15')),
 (8, 'Pool Table', Decimal('0'), Decimal('5'), Decimal('400'), Decimal('15'))]

So from the facilities table above, we have the data of facid as facility ID, name as a facility name, member cost, guest cost, initial outlay and monthly maintenance.

### 2. You want to print out a list of all of the facilities and their cost to members. How would you retrieve a list of only facility names and costs?

Print out a list of all of the facilities and their cost to members, but retrieve a list of only facility names and costs. We want to get the list of only facility names and the costs. 

In [7]:
%%sql 

SELECT name, membercost FROM cd.facilities;

 * postgresql://postgres:***@localhost/exercises
9 rows affected.


[('Tennis Court 1', Decimal('5')),
 ('Tennis Court 2', Decimal('5')),
 ('Badminton Court', Decimal('0')),
 ('Table Tennis', Decimal('0')),
 ('Massage Room 1', Decimal('35')),
 ('Massage Room 2', Decimal('35')),
 ('Squash Court', Decimal('3.5')),
 ('Snooker Table', Decimal('0')),
 ('Pool Table', Decimal('0'))]

So we see there are various facilities, like tennis courts, badminton court, massage rooms, squash court, snooker table, pool table and as well as how much they cost for the members.

### 3. How can you produce a list of facilities that charge a fee to members?

Produce a list of facilities that charge a fee to members. If we take a look back at the results from the previous question, note that not all of them actually charge a fee. Some of them have a zero cost for members, which means we want to go ahead and add a 'where' statement which filters out member costs that are basically zero.

In [8]:
 %%sql
  
SELECT * FROM cd.facilities 
WHERE membercost > 0;

 * postgresql://postgres:***@localhost/exercises
5 rows affected.


[(0, 'Tennis Court 1', Decimal('5'), Decimal('25'), Decimal('10000'), Decimal('200')),
 (1, 'Tennis Court 2', Decimal('5'), Decimal('25'), Decimal('8000'), Decimal('200')),
 (4, 'Massage Room 1', Decimal('35'), Decimal('80'), Decimal('4000'), Decimal('3000')),
 (5, 'Massage Room 2', Decimal('35'), Decimal('80'), Decimal('4000'), Decimal('3000')),
 (6, 'Squash Court', Decimal('3.5'), Decimal('17.5'), Decimal('5000'), Decimal('80'))]

We get back those five facilities where there's actually a cost to the member. Those facilites are Tennis Court 1 and 2, Massage Room 1 and 2, and Squash Court.

### 4. How can you produce a list of facilities that charge a fee to members, and that fee is less than 1/50th of the monthly maintenance cost? Return the facid, facility name, member cost, and monthly maintenance of the facilities in question.

Produce a list of facilities that charge a fee to members, and that fee is less than 1/50 of the monthly maintenance cost. Return the facid, facility name, member cost, and monthly maintenance of the facilities. So what we're looking for is for the membercost fee to be less than 1/50 of the monthly maintenance cost, or we can say we want to compare where that member cost is less than Monthly maintenance divided by 50.

In [9]:
 %%sql

SELECT facid, name, membercost, monthlymaintenance
FROM cd.facilities
WHERE membercost > 0 AND
(membercost < monthlymaintenance/50.0);

 * postgresql://postgres:***@localhost/exercises
2 rows affected.


[(4, 'Massage Room 1', Decimal('35'), Decimal('3000')),
 (5, 'Massage Room 2', Decimal('35'), Decimal('3000'))]

And there should actually just be two rows, Massage Room 1 and Massage Room 2.

### 5. How can you produce a list of all facilities with the word 'Tennis' in their name?

Produce a list of all facilities with the word 'Tennis' in their name.

In [10]:
%%sql

SELECT * FROM cd.facilities 
WHERE name LIKE '%Tennis%';

 * postgresql://postgres:***@localhost/exercises
3 rows affected.


[(0, 'Tennis Court 1', Decimal('5'), Decimal('25'), Decimal('10000'), Decimal('200')),
 (1, 'Tennis Court 2', Decimal('5'), Decimal('25'), Decimal('8000'), Decimal('200')),
 (3, 'Table Tennis', Decimal('0'), Decimal('5'), Decimal('320'), Decimal('10'))]

There are 3 results of facilities with the word 'Tennis' in it, which are Tennis Court 1, Tennis Court 2 and Table Tennis.

### 6. How can you retrieve the details of facilities with ID 1 and 5? Try to do it without using the OR operator.

Next is to retrieve the details of facilities of ID 1 and 5, and try to do it without using the OR operator.

In [11]:
%%sql

SELECT * FROM cd.facilities 
WHERE facid IN (1,5);

 * postgresql://postgres:***@localhost/exercises
2 rows affected.


[(1, 'Tennis Court 2', Decimal('5'), Decimal('25'), Decimal('8000'), Decimal('200')),
 (5, 'Massage Room 2', Decimal('35'), Decimal('80'), Decimal('4000'), Decimal('3000'))]

To do it without OR is by using the IN operator. What's nice about this IN operator is you can easily expand to any number of IDs by just adding it into the parentheses list instead of having to add a complete or statement again. So we get 2 rows results, facility ID 1 is tennis Court 1 and facility ID 5 is Massage Room 2.

### 7. How can you produce a list of members who joined after the start of September 2012? Return the memid, surname, firstname, and joindate of the members in question.

Produce a list of members who joined after the start of September 2012. Return the memid, surname, firstname, and joindate of the members. To get the join date members data we use members table.

In [12]:
%%sql

SELECT memid, surname, firstname, joindate 
FROM cd.members WHERE joindate >= '2012-09-01';

 * postgresql://postgres:***@localhost/exercises
10 rows affected.


[(24, 'Sarwin', 'Ramnaresh', datetime.datetime(2012, 9, 1, 8, 44, 42)),
 (26, 'Jones', 'Douglas', datetime.datetime(2012, 9, 2, 18, 43, 5)),
 (27, 'Rumney', 'Henrietta', datetime.datetime(2012, 9, 5, 8, 42, 35)),
 (28, 'Farrell', 'David', datetime.datetime(2012, 9, 15, 8, 22, 5)),
 (29, 'Worthington-Smyth', 'Henry', datetime.datetime(2012, 9, 17, 12, 27, 15)),
 (30, 'Purview', 'Millicent', datetime.datetime(2012, 9, 18, 19, 4, 1)),
 (33, 'Tupperware', 'Hyacinth', datetime.datetime(2012, 9, 18, 19, 32, 5)),
 (35, 'Hunt', 'John', datetime.datetime(2012, 9, 19, 11, 32, 45)),
 (36, 'Crumpet', 'Erica', datetime.datetime(2012, 9, 22, 8, 36, 38)),
 (37, 'Smith', 'Darren', datetime.datetime(2012, 9, 26, 18, 8, 45))]

So we have the results of member id, surname, first name, and join date of the members. There are of 10 members that joined either on or after September 1st, 2012. 

### 8. How can you produce an ordered list of the first 10 surnames in the members table? The list must not contain duplicates.

Produce an ordered list of the first 10 surnames in the members table. The list must not contain duplicates. So basically only the unique surnames will be shown as the result, and then to limit it to just ten.

In [13]:
%%sql

SELECT DISTINCT surname FROM cd.members
ORDER BY  surname LIMIT 10;

 * postgresql://postgres:***@localhost/exercises
10 rows affected.


[('Bader',),
 ('Baker',),
 ('Boothe',),
 ('Butters',),
 ('Coplin',),
 ('Crumpet',),
 ('Dare',),
 ('Farrell',),
 ('Genting',),
 ('GUEST',)]

We get the results, and the first one is in alphabetical order.

### 9. You'd like to get the signup date of your last member. How can you retrieve this information?

Get the signup date of your last member. Essentially, When did the last person sign up?

In [14]:
%%sql

SELECT MAX(joindate) AS latest_signup FROM cd.members;

 * postgresql://postgres:***@localhost/exercises
1 rows affected.


[(datetime.datetime(2012, 9, 26, 18, 8, 45),)]

Using the 'MAX' function, basically will give us the most recent date of the person who have signed up.

### 10. Produce a count of the number of facilities that have a cost to guests of 10 or more.

Produce a count of the number of facilities that have a cost to guests of 10 or more. We can get the guest cost from the facilities table. 

In [15]:
%%sql

SELECT COUNT(*) FROM cd.facilities 
WHERE guestcost >= 10;

 * postgresql://postgres:***@localhost/exercises
1 rows affected.


[(6,)]

There's six facilities that have a cost to guest equal to 10 or more. 

### 11. Produce a list of the total number of slots booked per facility in the month of September 2012. Produce an output table consisting of facility id and slots, sorted by the number of slots.

The booked data is in the bookings table. So we want to produce a list of the total number of slots booked per facility in the month of September 2012.

In [16]:
%%sql

SELECT facid, sum(slots) AS "Total Slots" FROM cd.bookings 
WHERE starttime >= '2012-09-01' AND starttime < '2012-10-01' 
GROUP BY facid 
ORDER BY SUM(slots);

 * postgresql://postgres:***@localhost/exercises
9 rows affected.


[(5, 122),
 (3, 422),
 (7, 426),
 (8, 471),
 (6, 540),
 (2, 570),
 (1, 588),
 (0, 591),
 (4, 648)]

To get the result of the total number of slots booked per facility in the month of September 2012. We can set the start time from 2012-09-01 untill 2012-10-01. So we get the result of the total slots per facility. Facility that has facility id number 5 with 122 total slots is the least booked whereas the facility id number 4 with 648 total slots is the most booked.

### 12. Produce a list of facilities with more than 1000 slots booked. Produce an output table consisting of facility id and total slots, sorted by facility id.

This is actually going to look extremely similar with the previous question, except now we want to filter by the some slots.

In [17]:
 %%sql

SELECT facid, sum(slots) AS total_slots FROM cd.bookings 
GROUP BY facid 
HAVING SUM(slots) > 1000 
ORDER BY facid;

 * postgresql://postgres:***@localhost/exercises
5 rows affected.


[(0, 1320), (1, 1278), (2, 1209), (4, 1404), (6, 1104)]

Using 'HAVING' because I'm filtering on some slots, so we'll say having the sum of the slots greater than 1000.
We get the result, there are 5 list of facilities with more than 1000 slots booked.

### 13. How can you produce a list of the start times for bookings for tennis courts, for the date '2012-09-21'? Return a list of start time and facility name pairings, ordered by the time.

Produce a list of the start times for bookings for tennis courts, for the date '2012-09-21'. Return a list of start time and facility name pairings, ordered by the time. We want to return a list of the start time and the facility name pairings ordered by the time. Also that we only want it for the tennis courts and we want their specific start time.

In [18]:
 %%sql

SELECT cd.bookings.starttime AS start, cd.facilities.name 
AS name 
FROM cd.facilities 
INNER JOIN cd.bookings
ON cd.facilities.facid = cd.bookings.facid 
WHERE cd.facilities.facid IN (0,1) 
AND cd.bookings.starttime >= '2012-09-21' 
AND cd.bookings.starttime < '2012-09-22' 
ORDER BY cd.bookings.starttime;

 * postgresql://postgres:***@localhost/exercises
12 rows affected.


[(datetime.datetime(2012, 9, 21, 8, 0), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 8, 0), 'Tennis Court 2'),
 (datetime.datetime(2012, 9, 21, 9, 30), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 10, 0), 'Tennis Court 2'),
 (datetime.datetime(2012, 9, 21, 11, 30), 'Tennis Court 2'),
 (datetime.datetime(2012, 9, 21, 12, 0), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 13, 30), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 14, 0), 'Tennis Court 2'),
 (datetime.datetime(2012, 9, 21, 15, 30), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 16, 0), 'Tennis Court 2'),
 (datetime.datetime(2012, 9, 21, 17, 0), 'Tennis Court 1'),
 (datetime.datetime(2012, 9, 21, 18, 0), 'Tennis Court 2')]

The bookings table have the booking ID, facility ID and a start time data. So we have a facility ID, but we don't actually know whether it's associated with a tennis court or not. In the facilities table we have the data the name of the facilities like Tennis Court 1 and Tennis Court 2 and their associated facility ID. So we can either use their name or their facility ID to filter out just the tennis courts. And then we need to join this using 'INNER JOIN' with that booking table in order to actually grab the start time. 'INNER JOIN' produces only the set of the records that match in both table bookings and table facilities. Next we can use the 'WHERE' and 'ORDER BY' statments.

We get the result, there are 12 rows that booking the Tennis Courts on 2012-09-21 date.

### 14. How can you produce a list of the start times for bookings by members named 'David Farrell'?

Produce a list of the start times for bookings by members named 'David Farrell'. We're going to do an 'INNER JOIN' between the members table and bookings table, because the members table actually hold the names and the bookings table holds the booking information. We only want booking information from the member named 'David Farrrel'.

In [19]:
 %%sql
    
SELECT cd.bookings.starttime 
FROM cd.bookings 
INNER JOIN cd.members ON 
cd.members.memid = cd.bookings.memid 
WHERE cd.members.firstname='David' 
AND cd.members.surname='Farrell';

 * postgresql://postgres:***@localhost/exercises
34 rows affected.


[(datetime.datetime(2012, 9, 18, 9, 0),),
 (datetime.datetime(2012, 9, 18, 13, 30),),
 (datetime.datetime(2012, 9, 18, 17, 30),),
 (datetime.datetime(2012, 9, 18, 20, 0),),
 (datetime.datetime(2012, 9, 19, 9, 30),),
 (datetime.datetime(2012, 9, 19, 12, 0),),
 (datetime.datetime(2012, 9, 19, 15, 0),),
 (datetime.datetime(2012, 9, 20, 11, 30),),
 (datetime.datetime(2012, 9, 20, 14, 0),),
 (datetime.datetime(2012, 9, 20, 15, 30),),
 (datetime.datetime(2012, 9, 21, 10, 30),),
 (datetime.datetime(2012, 9, 21, 14, 0),),
 (datetime.datetime(2012, 9, 22, 8, 30),),
 (datetime.datetime(2012, 9, 22, 17, 0),),
 (datetime.datetime(2012, 9, 23, 8, 30),),
 (datetime.datetime(2012, 9, 23, 17, 30),),
 (datetime.datetime(2012, 9, 23, 19, 0),),
 (datetime.datetime(2012, 9, 24, 8, 0),),
 (datetime.datetime(2012, 9, 24, 12, 30),),
 (datetime.datetime(2012, 9, 24, 16, 30),),
 (datetime.datetime(2012, 9, 25, 15, 30),),
 (datetime.datetime(2012, 9, 25, 17, 0),),
 (datetime.datetime(2012, 9, 26, 13, 0),),
 (datetime.datetime(2012, 9, 26, 17, 0),),
 (datetime.datetime(2012, 9, 27, 8, 0),),
 (datetime.datetime(2012, 9, 28, 9, 30),),
 (datetime.datetime(2012, 9, 28, 11, 30),),
 (datetime.datetime(2012, 9, 28, 13, 0),),
 (datetime.datetime(2012, 9, 29, 10, 30),),
 (datetime.datetime(2012, 9, 29, 13, 30),),
 (datetime.datetime(2012, 9, 29, 14, 30),),
 (datetime.datetime(2012, 9, 29, 16, 0),),
 (datetime.datetime(2012, 9, 29, 17, 30),),
 (datetime.datetime(2012, 9, 30, 14, 30),)]

So here's list of the start times for bookings by members named 'David Farrell'.